internal/application/origin.go

```diff
@@ -110,31 +113,58 @@ func SetOrigin(ctx context.Context, cluster *kubernetes.Cluster, app models.AppR
 		return err
 	}
 
-	// Assemble new origin json
-	var value string
+	patch, err := buildBodyPatch(origin)
+	if err != nil {
+		return errors.Wrap(err, "error building body patch")
+	}
+
+	_, err = client.Namespace(app.Namespace).Patch(ctx,
+		app.Name,
+		types.JSONPatchType,
+		patch,
+		metav1.PatchOptions{})
+
+	return err
+}
+
+func buildBodyPatch(origin models.ApplicationOrigin) ([]byte, error) {
+	patchOpValue := ApplicationOrigin{}
+
 	switch origin.Kind {
-	case models.OriginNone:
-		value = ""
+	case models.OriginNone: // do nothing (keeping here for reference)
 	case models.OriginPath:
-		value = fmt.Sprintf(`"path": "%s"`, origin.Path)
+		patchOpValue.Path = origin.Path
 	case models.OriginContainer:
-		value = fmt.Sprintf(`"container": "%s"`, origin.Container)
+		patchOpValue.Container = origin.Container
 	case models.OriginGit:
-		value = fmt.Sprintf(`"repository": "%s"`, origin.Git.URL)
+		patchOpValue.GitOrigin = &GitOrigin{Repository: origin.Git.URL}
 		if origin.Git.Revision != "" {
-			value = fmt.Sprintf(`%s, "revision": "%s"`, value, origin.Git.Revision)
+			patchOpValue.GitOrigin.Revision = origin.Git.Revision
 		}
-		value = fmt.Sprintf(`"git": {%s}`, value)
 	}
 
-	// And enter into the app
-	patch := fmt.Sprintf(`[{"op": "replace", "path": "/spec/origin", "value": {%s}}]`, value)
+	operations := []PatchOperation{{
+		Op:    "replace",
+		Path:  "/spec/origin",
+		Value: patchOpValue,
```

Comment:
Hm ... `patchOpValue` is the same type as `origin`, right ?
Could we use `origin` directly here ?
The preceding lines are effectively a more complicated `patchOpValue := origin` also ?! I.e could be dropped ?!

Reply:
Oh, I see now. You created a local copy of the model type, and all the code above is doing the copying.
Does the local copy have a different tagging for json marhsalling ? ...
Yes, the `Kind` field is not present in the local copy.
Ok then, retracting my thoughts above.

Reply:
Yes, I was thinking to use the same but unfortunately the GitOrigin is slightly different (`repository` vs `url`), not sure why.

Reply:
Please bring it up in the planning. I would like to have the code simple(r) here.
It may affect the server API, i.e. IIRC the model goes over the wire between client and server.
So a change would also affect @richard-cox

Reply:
As discussed I'm going to change the GitRef struct model to match the CRD

https://github.com/epinio/epinio/blob/dfc3aba919800f30a6a263949b9e7a813177359e/pkg/api/core/v1/models/app.go#L20-L23

https://github.com/epinio/helm-charts/blob/a3d66d6c0dda2d8c5752247f93a3e3fdf79064ec/chart/epinio/templates/app-crd.yaml#L59-L67

to

```golang
type GitRef struct {
	Revision   string `json:"revision,omitempty" yaml:"revision,omitempty"`
	Repository string `json:"repository"         yaml:"repository"`
}
```

I am going to update also the yaml, is that fine?

Reply:
I am not sure right now ... The yaml may have to match the CRD spec. I would not touch it, in the interest of making a minimal change to get the desired effect.

Note that you only have to change the `json:"url"` to ``json:"repository"`.
There is no need to change the field name in Go.
I.e. the `URL` to `repository` change is superfluous as well.

Reply:
Ok, let's keep it at the minimum then. :)